src/Features/Core/Portable/Formatting/IFormattingInterationService.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Formatting
+{
+    internal interface IFormattingInteractionService : ILanguageService
+    {
+        bool SupportsFormatDocument { get; }
+        bool SupportsFormatSelection { get; }
+        bool SupportsFormatOnPaste { get; }
+        bool SupportsFormatOnReturn { get; }
+
+        /// <summary>
+        /// True if this service would like to format the document based on the user typing the
+        /// provided character.
+        /// </summary>
+        bool SupportsFormattingOnTypedCharacter(Document document, char ch);
+
+        /// <summary>
+        /// Returns the text changes necessary to format the document.  If "textSpan" is provided,
+        /// only the text changes necessary to format that span are needed.
+        /// </summary>
+        Task<IList<TextChange>> GetFormattingChangesAsync(Document document, TextSpan? textSpan, DocumentOptionSet? documentOptions, CancellationToken cancellationToken);
```

Comment:
grumble grumble, i'm going to be a PITA, but we should have these be ImmutalbeArray, and i am ok if that means converting some true lists to be ImmutableArrays.  Thanks!